evaluate hand for pair, 2 pair, {3,4} of a kind, full house

andy5995 opened the pull request.

andy5995 committed:
evaluate hand for pair, 2 pair, {3,4} of a kind, full house

andy5995 committed:
remove commented code

Jammyjamjamman commented on src/lib.rs:
```diff
@@ -80,33 +80,116 @@ impl PokerRank {
+//
+pub fn evaluate(hand: &mut [Card; 5]) -> i32 {
+    hand.sort();
+    let last_index = hand.len() - 1;
```
We don't actually need this, because you've already forced the hand length to be 5 in the args. But this is fine.

Jammyjamjamman commented on src/lib.rs:
```diff
@@ -80,33 +80,116 @@ impl PokerRank {
+// Evalute for a poker hand.
+//
+//
+pub fn evaluate(hand: &mut [Card; 5]) -> i32 {
```
It might be better if the function returned the enum rather than an integer? (Or more Rust-style.)

andy5995 replied:
> It might be better if the function returned the enum rather than an integer? (Or more Rust-style.)

Yeah, it'll ultimately be something else. As we discussed, to break ties, we need to know some specifics about the ranks, etc.

Jammyjamjamman reviewed (commented):
Code looks fine. I'd probably define the evaluate() function as returning a PokerHandOrder enum, but it's fine if you're happier keeping it like this. I suppose you'd also have to add a NoMatch (-1) ad InvalidHand (-100) struct to the enum.

andy5995 committed:
add the "Nothing" poker rank

andy5995 commented:
>I suppose you'd also have to add a NoMatch (-1) ad InvalidHand (-100) struct to the enum.

I'll add "Nothing" to the PokerHandRank enum type, because that's what it's called in Poker. "I got Nothing." is the common colloquialism.

As for "InvalidHand", I'm not sure it's worth it to add that to the enum. It really shouldn't happen unless we introduce a program bug. Rust just complained because I didn't account for all cases in the match statement.

We could add a separate enum for custom error codes and strings? Or wrap the result in an io::Result and do something like this? https://github.com/theimpossibleastronaut/rmwrs/blob/fabcf801a65a7d86a380573cf60ef7dff6d85511/src/lib.rs#L139

andy5995 committed:
Complete first draft of poker hand evaluator

Now checks for all poker hand ranks

andy5995 committed:
move tests to separate module

andy5995 committed:
add (ignored) test that evaluates 100000 hands

---- tests::test_multiple_ranks stdout ----
Out of 1000000 hands dealt...

Pairs = 422267
Three of a kind = 7015
Straights = 3946
Straight Flushes = 11
Royal Flushes = 0

andy5995 committed:
publicize poker module
